text/0000-derive-bound-control.md

```diff
@@ -0,0 +1,831 @@
+apply to all derived traits.
+
+#### An example
+
+Given the following type definition:
+
+```rust
+#[derive(Clone)]
+struct Foo<#[derive_no_bound] S, T> {
+    bar: Arc<S>,
+    baz: T,
+}
+```
+
+The generated `impl` is:
+
+```rust
+impl<S, T: Clone> // <-- S: Clone is missing
+Clone for Foo { /* .. */ }
+```
+
+### Semantics - on a type
+
+When `#[derive_no_bound(..)]` is applied directly on a type, this is equivalent
+to specifying the identical attribute on each formal type parameter of the type.
+
+#### An example
+
+Consider a `Refl` encoding in Rust:
+
+```rust
+use std::marker::PhantomData;
+
+#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
+#[derive_no_bound]
+pub struct Id<S: ?Sized, T: ?Sized>(PhantomData<(*mut S, *mut T)>);
+```
+
+The generated `impl`s are:
+
+```rust
+impl<S: ?Sized, T: ?Sized> Copy       for Id<S, T> { /* .. */ }
+impl<S: ?Sized, T: ?Sized> Clone      for Id<S, T> { /* .. */ }
+impl<S: ?Sized, T: ?Sized> Debug      for Id<S, T> { /* .. */ }
+impl<S: ?Sized, T: ?Sized> Hash       for Id<S, T> { /* .. */ }
+impl<S: ?Sized, T: ?Sized> PartialEq  for Id<S, T> { /* .. */ }
+impl<S: ?Sized, T: ?Sized> Eq         for Id<S, T> { /* .. */ }
+impl<S: ?Sized, T: ?Sized> PartialOrd for Id<S, T> { /* .. */ }
+impl<S: ?Sized, T: ?Sized> Ord        for Id<S, T> { /* .. */ }
+```
+
+## `#[derive_field_bound(..)]`
+
+### Grammar
+
+The attribute `#[derive_field_bound(..)]` can be placed on type definitions
+directly (`struct`, `enum`, `union`) or on their fields. Note in particular
+that they may not be specified on variants of `enum`s. The attribute has the
+following grammar:
+
+```enbf
+field_bound_attr : "#" "[" "derive_field_bound" field_bound_traits? "]" ;
+field_bound_traits : "(" trait_list ","? ")" ;
+trait_list : ident | ident "," trait_list ;
+```
+
+### Semantics - on a field
+
+Formally: Assuming a field `F`, either named or unnamed, of
+type `FieldType`, and the attribute `#[derive_field_bound(Trait)]`
+on `F` for a specific trait`Trait`, specifying the attribute
+`#[derive(Trait)]` shall add a bound `FieldType: Trait` in the
+`where`-clause in the `impl<..> Trait<..> for Type<..>` generated
+by a derive macro for `Trait`.
+
+When `#[derive_field_bound(..)]` contains a comma separated list of traits,
+these semantics will apply to each trait referenced but not other traits.
+
+When `#[derive_field_bound]` is used (with no traits referenced), these rules
+will apply to all derived traits.
+
+#### An example
+
+Given the following type definition:
+
+```rust
+#[derive(Clone, PartialEq, PartialOrd)]
+struct Foo<S, T> {
+    #[derive_field_bound(Clone)]
+    bar: Bar<S>,
+    baz: Baz<T>
+}
+```
+
+The generated `impl`s are:
+
+```rust
+impl<S: Clone, T: Clone> Clone for Foo<S, T>
+where Bar<S>: Clone { /* .. */ } // <-- Note the where clause!
+
+impl<S: PartialEq, T: PartialEq> Clone for Foo<S, T> { /* .. */ }
+
+impl<S: PartialOrd, T: PartialEq> Clone for Foo<S, T> { /* .. */ }
```

Comment:
I think there are some copy-errors here (T: PartialEq), and all of the impls are for Clone

Reply:
Thanks for spotting these <3